### ISSUE TITLE
Using Sentry SDK in worker leads to crash

### ISSUE THREAD

**Issue by itssimon** [taskiq-python/taskiq#261]
The Taskiq workers crash immediately if I initialise the Sentry SDK in the worker startup handler.

```
$ taskiq worker --ack-type when_executed --reload apitally.tasks:broker                                                                4s  cloud  base 20:37:47
[2023-12-11 20:40:48,284][taskiq.worker][INFO   ][MainProcess] Starting 2 worker processes.
[2023-12-11 20:40:48,284][root][WARNING][MainProcess] Reload on change enabled. Number of worker processes set to 1.
[2023-12-11 20:40:48,290][taskiq.process-manager][INFO   ][MainProcess] Started process worker-0 with pid 30331 
objc[30331]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[30331]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[2023-12-11 20:40:50,301][taskiq.process-manager][INFO   ][MainProcess] worker-0 is dead. Scheduling reload.
[2023-12-11 20:40:51,314][taskiq.process-manager][INFO   ][MainProcess] Process worker-0 restarted with pid 30338
objc[30338]: +[NSNumber initialize] may have been in progress in another thread when fork() was called.
objc[30338]: +[NSNumber initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[2023-12-11 20:40:53,330][taskiq.process-manager][INFO   ][MainProcess] worker-0 is dead. Scheduling reload.
[2023-12-11 20:40:54,338][taskiq.process-manager][INFO   ][MainProcess] Process worker-0 restarted with pid 30342
objc[30342]: +[NSNumber initialize] may have been in progress in another thread when fork() was called.
objc[30342]: +[NSNumber initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

The worker startup handler looks like this:

```python
@broker.on_event(TaskiqEvents.WORKER_STARTUP)
async def on_worker_startup(state: TaskiqState) -> None:
    sentry_sdk.init(
        dsn=str(settings.sentry_dsn),
        environment=settings.sentry_environment,
    )
```

This doesn't happen if I don't use Sentry, or if I start the worker without the `--reload` flag.

Environment:
- macOS 14.0 (Sonoma)
- Python 3.11
- taskiq 0.10.4
- sentry-sdk 1.38.0

**Comment by itssimon** [comment]
Setting that environment variable doesn't work unfortunately, but the log output changes to:

```
[2023-12-11 21:25:53,237][taskiq.worker][INFO   ][MainProcess] Starting 2 worker processes.
[2023-12-11 21:25:53,237][root][WARNING][MainProcess] Reload on change enabled. Number of worker processes set to 1.
[2023-12-11 21:25:53,242][taskiq.process-manager][INFO   ][MainProcess] Started process worker-0 with pid 32423 
[2023-12-11 21:25:53,296][taskiq.worker][INFO   ][worker-0] Importing tasks from module apitally.tasks.tasks
The process has forked and you cannot use this CoreFoundation functionality safely. You MUST exec().
Break on __THE_PROCESS_HAS_FORKED_AND_YOU_CANNOT_USE_THIS_COREFOUNDATION_FUNCTIONALITY___YOU_MUST_EXEC__() to debug.
[2023-12-11 21:25:54,246][taskiq.process-manager][INFO   ][MainProcess] worker-0 is dead. Scheduling reload.
```

Even though this error is clearly rooted somewhere in macOS there must be a workaround. Taskiq is the only framework I've encountered this with, and I've been using hot reloading with FastAPI, Dramatiq, Celery, etc.

**Comment by s3rius** [comment]
I guess I have a possible solution. I will publish it soon. Will you be able to install taskiq from the branch to verify that the problem is gone?

**Comment by itssimon** [comment]
Ah, nice one! Yes, no problem.

**Comment by s3rius** [comment]
The branch is called `feature/macos-fork`. Please verify that now there's no problem.

**Comment by itssimon** [comment]
It's still not working, but the error is different:

```
[2023-12-11 21:43:29,859][taskiq.worker][INFO   ][MainProcess] Starting 2 worker processes.
[2023-12-11 21:43:29,859][root][WARNING][MainProcess] Reload on change enabled. Number of worker processes set to 1.
[2023-12-11 21:43:29,868][taskiq.process-manager][INFO   ][MainProcess] Started process worker-0 with pid 33654 
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/homebrew/Cellar/python@3.11/3.11.6/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.6/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/spawn.py", line 132, in _main
    self = reduction.pickle.load(from_parent)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.6/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/synchronize.py", line 115, in __setstate__
    self._semlock = _multiprocessing.SemLock._rebuild(*state)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory
[2023-12-11 21:43:30,973][taskiq.process-manager][INFO   ][MainProcess] worker-0 is dead. Scheduling reload.
```

**Comment by s3rius** [comment]
Okay. That was initial error that we were avoiding when choosing fork instead of spawn for macos. 

Let me give you a small overview of what is happening:
Taskiq's master process creates a bunch of children processes. They don't know about each other, only master process knows about it's children. On the startup we check that all children processes have started correctly and all signals have been set for each child (Signals are used for graceful shutdowns ([code](https://github.com/taskiq-python/taskiq/blob/6bd15508bf5ddeb7faed99d19c86673e63077d35/taskiq/cli/worker/run.py#L91C2-L91C2))). We achieve this by using the `multiprocessing.Event`. We give a reference to the event from the master process to it's children processes ([code](https://github.com/taskiq-python/taskiq/blob/6bd15508bf5ddeb7faed99d19c86673e63077d35/taskiq/cli/worker/process_manager.py#L183))  and wait for all events to be set ([code](https://github.com/taskiq-python/taskiq/blob/6bd15508bf5ddeb7faed99d19c86673e63077d35/taskiq/cli/worker/process_manager.py#L97C12-L97C12)). When all events are set, that all signal handlers have been successfully set up and we can continue. After that we import a broker, run all startup functions and start listening.

Because we used to use fork method, the macos security considerations disallow spawning another fork from forked process (as I understood). Maybe because infinite forking might lead to the [fork-bomb](https://www.wikiwand.com/en/Fork_bomb) exploit or something. The initial problem we observed happened because sentry tries to spawns another process that sends all the data to the sentry using fork method and fails to spawn it because of macOS' securities. 

If we change the multiprocessing method to spawn, which is "safer" from macos perspective, then we observe another problem. Now processes cannot be started because of `FileNotFoundError`. The origin of this issue is that `spawn` method actually doesn't have any shared memory with parent process and all variables recreated from scratch. Which leads to an error, since the `multiprocess.Event` is no longer can be dereferenced and accessed by the child process and therefore cannot be set. (Not sure why this error doesn't happen on linux, tho). 

This is how I see the problem. I see one possible solution. It's to start using `spawn` method by default and use another mechanism of readiness instead of the `multiprocessing.Event` . I will try to fix this error, but I don't have macos to verify it. I will tag you on github when the solution is ready so you can test it out. Let's solve the issue together. 

Have a nice day.

**Comment by itssimon** [comment]
Thanks for the detailed explanation. That all makes sense. I'm very happy to help solve this issue!

By the way, it may be good to set up automated tests to also run on [GitHub Actions macOS runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners). That way you can make sure Taskiq continues to work on macOS without you having a Mac yourself.

**Comment by s3rius** [comment]
We already use it for most places. But CLI and process manager is one of parts that is missing coverage. I will try to add integration testing later. But it's really hard to verify and test.

**Comment by s3rius** [comment]
@itssimon, hi! At first I came up with quite crazy idea of having [info-server](https://github.com/taskiq-python/taskiq/commit/3a097c504d985f1d4edb51ecb9b83871eb9820f3), but I disliked the way it goes, so I just removed the event. Can you try running taskiq from the branch [feature/new-process-mgr](https://github.com/taskiq-python/taskiq/pull/266)? If it solves the issue, we might merge it, but with some additional changes.

**Comment by itssimon** [comment]
Thanks for looking into this further. I ran taskiq from that branch and unfortunately the error remains identical to the one first reported above.

```
[2023-12-17 21:25:01,096][taskiq.worker][INFO   ][MainProcess] Starting 2 worker processes.
[2023-12-17 21:25:01,096][root][WARNING][MainProcess] Reload on change enabled. Number of worker processes set to 1.
[2023-12-17 21:25:01,101][taskiq.process-manager][INFO   ][MainProcess] Started process worker-0 with pid 3737 
objc[3737]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[3737]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[2023-12-17 21:25:02,109][taskiq.process-manager][INFO   ][MainProcess] worker-0 is dead. Scheduling reload.
```

**Comment by itssimon** [comment]
_(Disregard my previous deleted comment.)_

Running taskiq from that feature branch does indeed resolve the issue.

**Comment by s3rius** [comment]
Great. I will publish it after some small updates .

**Comment by igoose1** [comment]
@s3rius, I guess this can be closed as #266 is merged and @itssimon confirmed it fixed a bug.

**Comment by s3rius** [comment]
Yes, true. Thanks for noticing.
